src/extensions/Brush.js

```diff
@@ -17,201 +15,124 @@ const BRUSH_ELEMENT_NAMES = [
   return {
-    x1: rootProps.xScale(props.xDomain[0]),
-    x2: rootProps.xScale(props.xDomain[1]),
-    y1: rootProps.yScale(props.yDomain[0]),
-    y2: rootProps.yScale(props.yDomain[1]),
+    x1: Math.max(x1, x),
+    x2: Math.min(x2, x + width),
+    y1: Math.max(y1, y),
+    y2: Math.min(y2, y + height),
   }
 }
 
-function boundsToDomain(bounds, childProps) {
-  const {rootProps} = childProps
+function domainToBounds(rootProps, xDomain, yDomain) {
+  const {xScale, yScale} = rootProps
+  const [x1, x2] = xDomain
+  const [y1, y2] = yDomain
+  return {x1: xScale(x1), x2: xScale(x2), y1: yScale(y1), y2: yScale(y2)}
+}
+
+function boundsToDomain(bounds, xScale, yScale) {
+  const {x1, x2, y1, y2} = bounds
   return {
-    xDomain: [rootProps.xScale.invert(bounds.x1), rootProps.xScale.invert(bounds.x2)],
-    yDomain: [rootProps.yScale.invert(bounds.y1), rootProps.yScale.invert(bounds.y2)],
+    xDomain: [xScale.invert(x1), xScale.invert(x2)],
+    yDomain: [yScale.invert(y1), yScale.invert(y2)],
   }
 }
 
-const getBrushData = props => ({
-  x1: props.xDomain[0],
-  x2: props.xDomain[1],
-  y1: props.yDomain[0],
-  y2: props.yDomain[1],
-})
-
-function updateBounds(props, childProps, partialBounds) {
-  const newBounds = reorder({...props.bounds, ...partialBounds})
-  if (isOutOfBounds(newBounds, childProps.rootProps.plotRect)) {
-    const constrainedBounds = constraintToPlotRect(newBounds, childProps)
-    props.onUpdate(boundsToDomain(constrainedBounds, childProps))
-  } else {
-    props.onUpdate(boundsToDomain(newBounds, childProps))
+export default class Brush extends React.Component {
+  static propTypes = {
+    children: PropTypes.node,
+    onUpdate: PropTypes.func.isRequired,
+    xDomain: PropTypes.array, // eslint-disable-line react/forbid-prop-types
+    yDomain: PropTypes.array, // eslint-disable-line react/forbid-prop-types
   }
-}
 
-function mouseDown(props, childProps) {
-  const brushElementName = childProps.renderDatum && childProps.renderDatum.name
-  if (!includes(BRUSH_ELEMENT_NAMES, brushElementName)) {
-    props.onState({
-      bounds: {
-        x1: childProps.localMouse.x,
-        y1: childProps.localMouse.y,
-      },
-      brushElementName,
-    })
-    props.onUpdate({xDomain: undefined, yDomain: undefined})
-  } else {
-    props.onState({
-      bounds: domainToBounds(props, childProps),
-      brushElementName,
-    })
+  static defaultProps = {
+    xDomain: [],
+    yDomain: [],
   }
-}
 
-function mouseDrag(props, childProps) {
-  if (props.brushElementName === 'brushesCenter') {
-    const bounds = domainToBounds(props, childProps)
-    const mBounds = {
-      x1: bounds.x1 - childProps.mouseDelta.x,
-      x2: bounds.x2 - childProps.mouseDelta.x,
-      y1: bounds.y1 - childProps.mouseDelta.y,
-      y2: bounds.y2 - childProps.mouseDelta.y,
-    }
-    if (isOutOfBounds(mBounds, childProps.rootProps.plotRect)) {
-      const constrainedBounds = constraintToPlotRect(bounds, childProps)
```

Comment:
Note to self: this is called with `bounds`, not `mBounds` (if the rect is dragged outside the chart, we shouldn't squish it down)